railgun/internal/proxy/clientgo_cached_proxy_resolver.go

```diff
@@ -216,18 +217,25 @@ func (p *clientgoCachedProxyResolver) startCacheServer() {
 				p.logger.Error(err, "object could not be updated in the cache and will be discarded")
 				break
 			}
+			p.backendNeedsSync = true
 		case cobj := <-p.del:
 			if err := p.cacheDelete(cobj); err != nil {
 				p.logger.Error(err, "object could not be deleted from the cache and will be discarded")
 				break
 			}
+			p.backendNeedsSync = true
 		case <-p.syncTicker.C:
+			if !p.enableReverseSync && !p.backendNeedsSync {
+				break
+			}
+
 			updateConfigSHA, err := p.kongUpdater(p.ctx, p.lastConfigSHA, p.cache, p.ingressClassName, p.deprecatedLogger, p.kongConfig, p.enableReverseSync)
```

Comment:
It feels weird that (synchronous) proxy updates are on the same thread as the cache updates.

Imagine Admin API calls being very slow - in that case, the ingress controller will be effectively locked up, and only have a fraction of a single CPU core available.

Reply:
Proxy cache updates (e.g. proxy.UpdateObject) are asynchronous, they use buffered channels:

https://github.com/Kong/kubernetes-ingress-controller/blob/main/railgun/internal/proxy/clientgo_cached_proxy_resolver.go#L70

The ingress controller (among other controllers that use the proxy cache) will only lock if the buffer reaches `DefaultObjectBufferSize` (currently set to 500 objects).

For a little bit more explanation:

If we ever reach that level somehow the behavior is to [throw an error](https://github.com/Kong/kubernetes-ingress-controller/blob/main/railgun/internal/proxy/clientgo_cached_proxy_resolver.go#L173):

```go
return fmt.Errorf("the proxy is too busy to accept requests at this time, try again later")
```

Which would then be [caught as an error during reconcilation](https://github.com/Kong/kubernetes-ingress-controller/blob/main/railgun/controllers/configuration/zz_generated_controllers.go#L161) and trigger a re-queue:

```go
if err := r.Proxy.UpdateObject(obj); err != nil {
	return ctrl.Result{}, err
}
```

So while yes, the `kongUpdater` calls ultimately do block the goroutine for the cache and can create back-pressure on the controllers that can lead to eventual blocking; this was an intentional design and tolerates an amount of backpressure defined in this documentation:

https://github.com/Kong/kubernetes-ingress-controller/blob/main/railgun/internal/proxy/proxy.go#L24

And moreover has failure mechanisms in place that are eventually consistent.

More to you point: could we improve the asynchronicity of the Kong Admin API updates and further desynchronize those updates from updates to the cache implementation? Yes, we could but my point in illustrating all the deliberate choices above is that in the current system we provide a back-pressure mechanism that couples the reconcilation of an object to the successful configuration of that object in the backend _intentionally_ at a specific threshold.

I hope the above clears up any confusion about ingress controllers being locked up. If you still feel strongly that we need further improvement I'm quite open to suggestions for alternative implementations and improvements that we can make to this functionality, however I do not think they belong in the scope of this small performance improvement, nor should they block them and I would want to take them into a separate issue so we can do some considerations before we take a stab at an alternative implementation.

Reply:
proxy.KongUpdater is called every syncPeriod, it is a bottleneck if we do not spin up dedicated goroutine. 

Reconcile that invokes proxy.UpdateObject (en object into cache) is more frequency, and I have been assuming they are coupled.

Reply:
No action required from my side - just wanted to point out this phenomenon, but if it's intentional then :+1: